src/job_token.rs

```diff
@@ -0,0 +1,70 @@
+use jobserver::{Acquired, Client, HelperThread};
+use std::sync::mpsc::{self, Receiver, Sender};
+
+pub(crate) struct JobToken {
+    /// The token can either be a fresh token obtained from the jobserver or - if `token` is None - an implicit token for this process.
+    /// Both are valid values to put into queue.
+    token: Option<Acquired>,
+    pool: Sender<Option<Acquired>>,
```

Comment:
You could make this an `Option` and remove `should_return_to_queue`:

```suggestion
    pool: Option<Sender<Option<Acquired>>>,
```

If `pool` is `Some(..)`, then send the `token` back to the `pool`.